rfcs/20231012-external-mel-bindings.md

```diff
@@ -0,0 +1,546 @@
+
+```ocaml
+type t
+external book : unit -> t = "Book" [@@mel.new] [@@mel.module]
+let myBook = book ()
+```
+
+After:
+
+```ocaml
+type t
+let%js.import book: t = "Book"
+let%js create_book : ~t_:t -> unit -> t = "new $t_()"
+let myBook = book |> create_book ()
+```
+
+## `mel.send` and `mel.send.pipe`
+
+Current:
+
+```ocaml
+external get_by_id : document -> string -> Dom.element = "getElementById"
+  [@@mel.send]
+external get_by_id : string -> Dom.element = "getElementById"
+  [@@mel.send.pipe: document]
+```
+
+Proposed:
+
+```ocaml
+let%js get_by_id : doc_:document -> id:string -> Dom.element = "$doc_.getElementById($id)"
+let%js get_by_id : id:string -> doc_:document -> Dom.element = "$doc_.getElementById($id)"
+```
+
+## `mel.new`
+
+Current:
+
+```ocaml
+type t
+external create_date : unit -> t = "Date" [@@mel.new]
+let date = create_date ()
+```
+
+Proposed:
+
+```ocaml
+type t
+let%js create_date : unit -> t = "new Date()"
+let date = create_date ()
+```
+
+## `mel.scope`
+
+Before:
+
+```ocaml
+external node_env : string = "NODE_ENV" [@@mel.scope "process", "env"]
+```
+
+Proposed:
+
+```ocaml
+let%js node_env : string = "process.env.NODE_ENV"
+```
+
+## `mel.uncurry` and `u`
+
+Uncurrying in Melange has historically been complicated, but there might be ways
+to make it simpler to understand and use.
+
+The need for uncurrying OCaml functions in Melange arises when both these
+situations happen:
+- A function written in OCaml is being passed to some JavaScript library through
+  the bindings
+- *And* the function can potentially be called from the JavaScript library
+  itself
+
+Sometimes only the first situation happens. For example, if we write bindings to
+`React.useCallback` we do not need to care about uncurrying, because the `react`
+JavaScript library itself will never call the callback, it will always be called
+from the OCaml side.
+
+To reflect this and help Melange users identify the scenarios where it's needed,
+a attribute `mel.called_from_js` is introduced.
+
+Before:
+
+```ocaml
+external map :
+  'a array -> 'b array -> (('a -> 'b -> 'c)[@mel.uncurry]) -> 'c array = "map"
+```
+
+After:
+
+```ocaml
+let%js map :
+    arr1:'a array ->
+    arr2:'b array ->
+    cb:(('a -> 'b -> 'c)[@mel.called_from_js]) ->
+    'c array =
+  {|map($arr1, $arr2, $cb)|}
+```
+
+When the PPX detects the usage of this attribute on arrow types in `let%js`
+extensions, it will transform the generated code so the function passed is
+transformed into its uncurried form, in a similar way than js_of_ocaml does with
+its runtime function
+[Js.wrap_callback](https://github.com/ocsigen/js_of_ocaml/blob/0fc2b735593f8b18bc5aa4342a3c79f4a39fd81f/lib/js_of_ocaml/js.mli#L160-L163).
+
+Using `mel.raw` to show how the example above would look like:
+
+```ocaml
+let map :
+    arr1:'a array ->
+    arr2:'b array ->
+    cb:('a -> 'b -> 'c) ->
+    'c array =
+  [%mel.raw
+    "function (arr1, arr2, cb) { return map(arr1, arr2, Js.wrap_callback(cb)) }"]
```

Comment:
what's the definition of `Js.wrap_callback(cb)`? I don't understand this example.

Reply:
https://github.com/ocsigen/js_of_ocaml/blob/3954ace2f4da41052d4e83b52f89bdb9ee0c3629/runtime/jslib.js#L306-L318